### PR TITLE
[UNR-838][MS] Adding connection saturation 

### PR DESCRIPTION
**Pull request by MatthewSandfordImprobable**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

These changes allow the tracking of the number of bytes we send when replicating actors. This is used to work out of the SpatialNetConnection is saturated. 

Update: This PR now only allows ReplicateActor to return the number of bits written.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 1]
```diff
@@ -743,7 +751,15 @@ int64 USpatialActorChannel::ReplicateActor()
...
 
-	return (bWroteSomethingImportant) ? 1 : 0;	// TODO: return number of bits written (UNR-664)
+	if (ReplicationBytesWritten > 0)
+	{
+		if (USpatialNetConnection* SpatialConnection = Cast<USpatialNetConnection>(Connection))
+		{
+			SpatialConnection->AddQueuedBytes(ReplicationBytesWritten);
+		}
+	}
+
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: driveby - can you add
`INC_DWORD_STAT_BY(STAT_NumReplicatedActorBytes, ReplicationBytesWritten);`

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 2]
```diff
@@ -743,7 +751,15 @@ int64 USpatialActorChannel::ReplicateActor()
...
+	return ReplicationBytesWritten;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: This should return bits written

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 3]
```diff
@@ -540,7 +540,8 @@ void USpatialReceiver::HandleActorAuthority(const Worker_AuthorityChangeOp& Op)
 			{
 				if (UObject* Object = PendingSubobjectAttachment.Subobject.Get())
 				{
-					Sender->SendAddComponent(PendingSubobjectAttachment.Channel, Object, *PendingSubobjectAttachment.Info);
+					uint32 BytesWritten = 0;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Can you add the TODO here also regarding saturation?

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Another instance where this object's replication will be desynced with the rest of the actor replication

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h [hunk 4]
```diff
@@ -310,6 +310,8 @@ class SPATIALGDK_API USpatialActorChannel : public UActorChannel
 
 	uint8 FramesTillDormancyAllowed = 0;
 
+	uint32 ReplicationBytesWritten = 0;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h]: Clarification that this isn't for the lifetime, but per frame

File: SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp [hunk 5]
```diff
@@ -330,29 +333,37 @@ void ComponentFactory::AddProperty(Schema_Object* Object, Schema_FieldId FieldId
 	}
 }
 
-TArray<FWorkerComponentData> ComponentFactory::CreateComponentDatas(UObject* Object, const FClassInfo& Info, const FRepChangeState& RepChangeState, const FHandoverChangeState& HandoverChangeState)
+TArray<FWorkerComponentData> ComponentFactory::CreateComponentDatas(UObject* Object, const FClassInfo& Info, const FRepChangeState& RepChangeState, const FHandoverChangeState& HandoverChangeState, uint32& OutBytesWritten)
 {
 	TArray<FWorkerComponentData> ComponentDatas;
 
+	OutBytesWritten = 0;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp]: Not sure we should do this

Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp]: Why not? All functions inside ComponentFactory set OutBytesWritten instead of adding to it.

**Comment by MatthewSandfordImprobable** [comment]
image

We track the time to replicate here. We can use this to track the impact of tracking bytes sent.

image

Additionally we have this Stat.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 6]
```diff
@@ -743,7 +752,9 @@ int64 USpatialActorChannel::ReplicateActor()
 
 	bForceCompareProperties = false;		// Only do this once per frame when set
 
-	return (bWroteSomethingImportant) ? 1 : 0;	// TODO: return number of bits written (UNR-664)
+	INC_DWORD_STAT_BY(STAT_SpatialNumReplicatedActorBytes, ReplicationBytesWritten);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: No need to create our own variable - lets just use `STAT_NumReplicatedActorBytes`

File: SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp (hunk 5 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp]: I think it's fine. I would expect that I get a set result rather than an additive result

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h (hunk 4 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h]: +1

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 7]
```diff
@@ -743,7 +752,9 @@ int64 USpatialActorChannel::ReplicateActor()
...
+
+	return ReplicationBytesWritten * 8;
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Isn't it bits if you're doing a * 8 :o

Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Yeah and that's what ReplicateActor returns.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp (hunk 5 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp]: Oh boy... you won't like my next commit.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 7 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Get in the game Josh

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 8]
```diff
@@ -414,7 +417,7 @@ FHandoverChangeState USpatialActorChannel::CreateInitialHandoverChangeState(cons
 
 int64 USpatialActorChannel::ReplicateActor()
 {
-	SCOPE_CYCLE_COUNTER(STAT_SpatialActorChannelReplicateActor);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: SCOPE_CYCLE_COUNTER should stay in to measure the time spent in this function.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 7 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: My point is, shouldn't the var be called `ReplicationBitsWritten` if we then have to convert it to bytes when we return by multiplying it by 8.  Also doesn't it mean the stat `STAT_SpatialNumReplicatedActorBytes` should be bits? Sorry if I'm missing something basic here.

Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: How many bits are in 2 bytes:
2 (Bytes) * 8 = 16 (Bits)

**Comment by MatthewSandfordImprobable** [comment]
Can see the bytes replicated and num actors replicated in trace. I tell you, spatial uses like 6x bandwidth :(

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 7 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Ok my idiot streak was a success here. Yeah I follow now - brain fart.
